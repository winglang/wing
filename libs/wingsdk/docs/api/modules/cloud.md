[@monadahq/wingsdk](../README.md) / [Exports](../modules.md) / cloud

# Namespace: cloud

## Table of contents

### Enumerations

- [BucketInflightMethods](../enums/cloud.BucketInflightMethods.md)
- [FunctionInflightMethods](../enums/cloud.FunctionInflightMethods.md)
- [QueueInflightMethods](../enums/cloud.QueueInflightMethods.md)

### Classes

- [Bucket](../classes/cloud.Bucket.md)
- [BucketBase](../classes/cloud.BucketBase.md)
- [Function](../classes/cloud.Function.md)
- [FunctionBase](../classes/cloud.FunctionBase.md)
- [Queue](../classes/cloud.Queue.md)
- [QueueBase](../classes/cloud.QueueBase.md)
- [Resource](../classes/cloud.Resource.md)

### Interfaces

- [BucketProps](../interfaces/cloud.BucketProps.md)
- [FunctionProps](../interfaces/cloud.FunctionProps.md)
- [IBucketClient](../interfaces/cloud.IBucketClient.md)
- [IFunctionClient](../interfaces/cloud.IFunctionClient.md)
- [IQueueClient](../interfaces/cloud.IQueueClient.md)
- [QueueOnMessageProps](../interfaces/cloud.QueueOnMessageProps.md)
- [QueueProps](../interfaces/cloud.QueueProps.md)

### Variables

- [BUCKET\_TYPE](cloud.md#bucket_type)
- [FUNCTION\_TYPE](cloud.md#function_type)
- [QUEUE\_TYPE](cloud.md#queue_type)

## Variables

### BUCKET\_TYPE

• `Const` **BUCKET\_TYPE**: ``"wingsdk.cloud.Bucket"``

Global identifier for `Bucket`.

#### Defined in

[src/cloud/bucket.ts:9](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/bucket.ts#L9)

___

### FUNCTION\_TYPE

• `Const` **FUNCTION\_TYPE**: ``"wingsdk.cloud.Function"``

Global identifier for `Function`.

#### Defined in

[src/cloud/function.ts:9](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/function.ts#L9)

___

### QUEUE\_TYPE

• `Const` **QUEUE\_TYPE**: ``"wingsdk.cloud.Queue"``

Global identifier for `Queue`.

#### Defined in

[src/cloud/queue.ts:10](https://github.com/monadahq/winglang/blob/438eedb/libs/wingsdk/src/cloud/queue.ts#L10)
