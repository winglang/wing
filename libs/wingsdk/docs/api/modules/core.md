[@monadahq/wingsdk](../README.md) / [Exports](../modules.md) / core

# Namespace: core

## Table of contents

### Enumerations

- [Language](../enums/core.Language.md)

### Classes

- [App](../classes/core.App.md)
- [Code](../classes/core.Code.md)
- [DependencyGraph](../classes/core.DependencyGraph.md)
- [DependencyVertex](../classes/core.DependencyVertex.md)
- [Duration](../classes/core.Duration.md)
- [Inflight](../classes/core.Inflight.md)
- [InflightClient](../classes/core.InflightClient.md)
- [NodeJsCode](../classes/core.NodeJsCode.md)
- [Synthesizer](../classes/core.Synthesizer.md)
- [Testing](../classes/core.Testing.md)

### Interfaces

- [AppProps](../interfaces/core.AppProps.md)
- [Capture](../interfaces/core.Capture.md)
- [CaptureMetadata](../interfaces/core.CaptureMetadata.md)
- [ICapturable](../interfaces/core.ICapturable.md)
- [ICapturableConstruct](../interfaces/core.ICapturableConstruct.md)
- [InflightBundleOptions](../interfaces/core.InflightBundleOptions.md)
- [InflightProps](../interfaces/core.InflightProps.md)
- [SynthesizerProps](../interfaces/core.SynthesizerProps.md)
