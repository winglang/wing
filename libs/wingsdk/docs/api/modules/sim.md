[@monadahq/wingsdk](../README.md) / [Exports](../modules.md) / sim

# Namespace: sim

## Table of contents

### Classes

- [App](../classes/sim.App.md)
- [Bucket](../classes/sim.Bucket.md)
- [BucketClient](../classes/sim.BucketClient.md)
- [Function](../classes/sim.Function.md)
- [FunctionClient](../classes/sim.FunctionClient.md)
- [PolyconFactory](../classes/sim.PolyconFactory.md)
- [Queue](../classes/sim.Queue.md)
- [QueueClient](../classes/sim.QueueClient.md)
- [Synthesizer](../classes/sim.Synthesizer.md)

### Interfaces

- [AppProps](../interfaces/sim.AppProps.md)
- [BaseResourceSchema](../interfaces/sim.BaseResourceSchema.md)
- [BucketSchema](../interfaces/sim.BucketSchema.md)
- [ConstructSchema](../interfaces/sim.ConstructSchema.md)
- [FunctionSchema](../interfaces/sim.FunctionSchema.md)
- [IResource](../interfaces/sim.IResource.md)
- [QueueSchema](../interfaces/sim.QueueSchema.md)
- [QueueSubscriber](../interfaces/sim.QueueSubscriber.md)
- [WingSimulatorSchema](../interfaces/sim.WingSimulatorSchema.md)

### Type Aliases

- [FunctionId](sim.md#functionid)

### Functions

- [isResource](sim.md#isresource)

## Type Aliases

### FunctionId

Ƭ **FunctionId**: `string`

#### Defined in

[src/sim/schema-resources.ts:4](https://github.com/monadahq/winglang/blob/main/libs/wingsdk/src/sim/schema-resources.ts#L4)

## Functions

### isResource

▸ **isResource**(`obj`): obj is IResource

#### Parameters

| Name | Type |
| :------ | :------ |
| `obj` | `any` |

#### Returns

obj is IResource

#### Defined in

[src/sim/resource.ts:12](https://github.com/monadahq/winglang/blob/main/libs/wingsdk/src/sim/resource.ts#L12)
