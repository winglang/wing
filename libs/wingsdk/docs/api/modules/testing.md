[@monadahq/wingsdk](../README.md) / [Exports](../modules.md) / testing

# Namespace: testing

## Table of contents

### Classes

- [Simulator](../classes/testing.Simulator.md)

### Interfaces

- [IResourceResolver](../interfaces/testing.IResourceResolver.md)
- [ISimulatorDispatcher](../interfaces/testing.ISimulatorDispatcher.md)
- [SimulatorContext](../interfaces/testing.SimulatorContext.md)
- [SimulatorProps](../interfaces/testing.SimulatorProps.md)
