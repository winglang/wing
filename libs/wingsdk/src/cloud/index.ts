export * from "./api";
export * from "./bucket";
export * from "./counter";
export * from "./domain";
export * from "./endpoint";
export * from "./function";
export * from "./on-deploy";
export * from "./queue";
export * from "./schedule";
export * from "./secret";
export * from "./service";
export * from "./topic";
export * from "./website";
