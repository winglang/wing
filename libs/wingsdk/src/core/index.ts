export * from "./app";
export * from "./attributes";
export * from "./connections";
export * from "./dependency";
export * from "./inflight";
export * from "./lifting";
export * from "./tree";
