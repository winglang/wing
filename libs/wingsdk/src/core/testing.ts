import { Code, NodeJsCode } from "./inflight";
import { PREBUNDLE_SYMBOL } from "./internal";
import { Resource } from "./resource";

/**
 * Testing utilities.
 */
export class Testing {
  /**
   * Obtain a reference to the prebundled Code for a given capture scope.
   */
  public static inspectPrebundledCode(captureScope: Resource): Code {
    const prebundle = (captureScope as any)[PREBUNDLE_SYMBOL];
    if (!prebundle) {
      throw new Error("No prebundled code found on this resource.");
    }

    // TODO This is a hack. Our path for inflight requires should be relative
    return NodeJsCode.fromInline(removeAbsolutePath(prebundle.text));
  }
}

function removeAbsolutePath(text: string) {
  const regex = /"\/.+?\/libs\/wingsdk\/(.+?)"/g;

  // replace first group with static text
  return text.replace(regex, '"[REDACTED]/wingsdk/$1"');
}
