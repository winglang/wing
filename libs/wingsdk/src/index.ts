// this file must only export the standard library types and not target-specific types.
export * as cloud from "./cloud";
export * as core from "./core";
export * as ex from "./ex";
export * as expect from "./expect";
export * as fs from "./fs";
export * as helpers from "./helpers";
export * as http from "./http";
export * as math from "./math";
export * as platform from "./platform";
export * as regex from "./regex";
export * as aws from "./shared-aws";
export * as simulator from "./simulator";
export * as std from "./std";
export * as sim from "./target-sim";
export * as ui from "./ui";
export * as util from "./util";
