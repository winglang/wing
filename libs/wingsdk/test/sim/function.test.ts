import * as cloud from "../../src/cloud";
import * as core from "../../src/core";
import * as sim from "../../src/sim";
import { IFunctionClient } from "../../src/sim";
import * as testing from "../../src/testing";
import { mkdtemp } from "../../src/util";
import { simulatorJsonOf } from "./util";

const INFLIGHT_CODE = core.NodeJsCode.fromInline(`
async function $proc($cap, event) {
  event = JSON.parse(event);
  let msg;
  if (process.env.PIG_LATIN) {
    msg = "Ellohay, " + event.name + "!";
  } else {
    msg = "Hello, " + event.name + "!";
  }
  return { msg };
}`);

test("create a function", async () => {
  // GIVEN
  const app = new sim.App({ outdir: mkdtemp() });
  const handler = new core.Inflight({
    code: INFLIGHT_CODE,
    entrypoint: "$proc",
  });
  new cloud.Function(app, "my_function", handler, {
    env: {
      ENV_VAR1: "true",
    },
  });
  const simfile = app.synth();

  // THEN
  const s = new testing.Simulator({ simfile });
  await s.start();
  expect(s.getAttributes("root/my_function")).toEqual({
    handle: expect.any(String),
  });
  expect(s.getProps("root/my_function")).toEqual({
    sourceCodeFile: expect.any(String),
    sourceCodeLanguage: "javascript",
    environmentVariables: {
      ENV_VAR1: "true",
      WING_SIM_RUNTIME_FUNCTION_PATH: "root/my_function",
    },
  });
  await s.stop();

  expect(simulatorJsonOf(simfile)).toMatchSnapshot();
});

test("invoke function", async () => {
  // GIVEN
  const app = new sim.App({ outdir: mkdtemp() });
  const handler = new core.Inflight({
    code: INFLIGHT_CODE,
    entrypoint: "$proc",
  });
  new cloud.Function(app, "my_function", handler);
  const simfile = app.synth();

  const s = new testing.Simulator({ simfile });
  await s.start();

  const client = s.getResourceByPath("root/my_function") as IFunctionClient;

  // WHEN
  const PAYLOAD = { name: "Alice" };
  const response = await client.invoke(JSON.stringify(PAYLOAD));

  // THEN
  expect(response).toEqual({ msg: `Hello, ${PAYLOAD.name}!` });
  await s.stop();

  expect(simulatorJsonOf(simfile)).toMatchSnapshot();
});

test("invoke function with environment variables", async () => {
  // GIVEN
  const app = new sim.App({ outdir: mkdtemp() });
  const handler = new core.Inflight({
    code: INFLIGHT_CODE,
    entrypoint: "$proc",
  });
  new cloud.Function(app, "my_function", handler, {
    env: {
      PIG_LATIN: "true",
    },
  });
  const simfile = app.synth();

  const s = new testing.Simulator({ simfile });
  await s.start();

  const client = s.getResourceByPath("root/my_function") as IFunctionClient;

  // WHEN
  const PAYLOAD = { name: "Alice" };
  const response = await client.invoke(JSON.stringify(PAYLOAD));

  // THEN
  expect(response).toEqual({
    msg: `Ellohay, ${PAYLOAD.name}!`,
  });
  await s.stop();

  expect(simulatorJsonOf(simfile)).toMatchSnapshot();
});
