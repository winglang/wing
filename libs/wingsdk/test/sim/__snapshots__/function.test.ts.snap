// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create a function 1`] = `
Object {
  "root": Object {
    "children": Object {
      "my_function": Object {
        "props": Object {
          "environmentVariables": Object {
            "ENV_VAR1": "true",
            "WING_SIM_RUNTIME_FUNCTION_PATH": "root/my_function",
          },
          "sourceCodeFile": "assets/my_function/index.js",
          "sourceCodeLanguage": "javascript",
        },
        "type": "wingsdk.cloud.Function",
      },
    },
    "type": "constructs.Construct",
  },
  "sdkVersion": "0.0.0",
  "startOrder": Array [
    "root/my_function",
  ],
}
`;

exports[`invoke function 1`] = `
Object {
  "root": Object {
    "children": Object {
      "my_function": Object {
        "props": Object {
          "environmentVariables": Object {
            "WING_SIM_RUNTIME_FUNCTION_PATH": "root/my_function",
          },
          "sourceCodeFile": "assets/my_function/index.js",
          "sourceCodeLanguage": "javascript",
        },
        "type": "wingsdk.cloud.Function",
      },
    },
    "type": "constructs.Construct",
  },
  "sdkVersion": "0.0.0",
  "startOrder": Array [
    "root/my_function",
  ],
}
`;

exports[`invoke function with environment variables 1`] = `
Object {
  "root": Object {
    "children": Object {
      "my_function": Object {
        "props": Object {
          "environmentVariables": Object {
            "PIG_LATIN": "true",
            "WING_SIM_RUNTIME_FUNCTION_PATH": "root/my_function",
          },
          "sourceCodeFile": "assets/my_function/index.js",
          "sourceCodeLanguage": "javascript",
        },
        "type": "wingsdk.cloud.Function",
      },
    },
    "type": "constructs.Construct",
  },
  "sdkVersion": "0.0.0",
  "startOrder": Array [
    "root/my_function",
  ],
}
`;
